spring-data-relational/src/main/java/org/springframework/data/relational/core/mapping/BasicRelationalPersistentEntity.java

```diff
@@ -37,16 +37,17 @@
  * @author Bastian Wilhelm
  * @author Mikhail Polivakha
  * @author Kurt Niemi
+ * @author Sergey Korotaev
  */
 class BasicRelationalPersistentEntity<T> extends BasicPersistentEntity<T, RelationalPersistentProperty>
 		implements RelationalPersistentEntity<T> {
 
 	private static final SpelExpressionParser PARSER = new SpelExpressionParser();
 
 	private final Lazy<SqlIdentifier> tableName;
-	private final @Nullable Expression tableNameExpression;
+	private final Lazy<SqlIdentifier> tableNameExpression;
```

Comment:
That's not what was meant here. We should remain with `Expression` (or switch to `ValueExpression`).

It was rather meant that a `Expression` (respective `ValueExpression`) should be able to return an `SqlIdentifier` and not be forced to return a `String` (as it is done now).

We also want to evaluate the expression each time we obtain names and we don't want to introduce caching.

Reply:
`ExpressionEvaluator` should return a `SqlIdentifier`. This is a package-protected class and we're free to change its signature. 

I suggest renaming it to `SqlIdentifierExpressionEvaluator` and not force a `String` type during expression evaluation (`expression.getValue(…)`), but introspect the returned object. 

If it is already a `SqlIdentifier`, then just return it as-is, otherwise, transform the value to `String` and apply `SqlIdentifierSanitizer`. Then, apply force-quoting as per `isForceQuote()` of the calling code.

Does that help or did that increase confusion?

Reply:
No, thanks, it's clear! In first time I thought about something like this (not so detailed) but I was scary and not sure to change `ExpressionEvaluator`, sorry....
I'll do it a little bit later